### PR TITLE
Create count negative numbers in matrix algorithm

### PR DESCRIPTION
**Pull request by CaedenPH**
### Describe your change:
Three implementations of this algorithm are implemented and benchmarked together with doctests

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by CaedenPH** [comment]
Note this pr supposedly has lack of tests however the benchmark function shouldn't have doctests anyway
https://github.com/TheAlgorithms/Python/blob/master/ciphers/atbash.py#L41
As shown, this benchmark function doesnt have doctests either

**Comment by cclauss** [comment]
When would we want to use this algorithm?

File: matrix/count_negative_numbers_in_sorted_matrix.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
+"""
+Given an `m x n` matrix grid which is sorted in non-increasing order
```
Comment (cclauss) [inline matrix/count_negative_numbers_in_sorted_matrix.py:2]:
```suggestion
Given an `m x n` matrix grid which is sorted in decreasing order
```

File: matrix/count_negative_numbers_in_sorted_matrix.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+both row-wise and column-wise, return the number of negative numbers in grid.
```
Comment (cclauss) [inline matrix/count_negative_numbers_in_sorted_matrix.py:3]:
```suggestion
both first row-wise and then column-wise, return the number of negative numbers in grid.
```

File: matrix/count_negative_numbers_in_sorted_matrix.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
+"""
+Given an `m x n` matrix grid which is sorted in decreasing order
+both row-wise and column-wise, return the number of negative numbers in grid.
```
Comment (cclauss) [inline matrix/count_negative_numbers_in_sorted_matrix.py:3]: Now I understand better...
```suggestion
Given an matrix of numbers in which all rows and all columns are sorted in decreasing
order, return the number of negative numbers in grid.
```

**Comment by CaedenPH** [comment]
> When would we want to use this algorithm?

@cclauss If you have a grid of sorted numbers in descending order in a matrix, this algorithm can be used to count the number of negative numbers. Not sure when someone would want to use it, but then again when would someone want to use the quine from other/quine?
As far as I'm aware this repository is just about algorithms, not necessarily useful ones. 

To answer your previous comment about validating the matrix, I don't think that is necessary. For searches that require sorted arrays, there aren't checks to ensure that those arrays are sorted so I think indicating that the matrix must be sorted is good enough

File: matrix/count_negative_numbers_in_sorted_matrix.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    0
+    >>> count_negatives_brute_force([[7,7,6],[-1,-2,-3]])
+    3
+    """
+    total = 0
+    for m in range(len(grid)):
+        for n in range(len(grid[m])):
+            if grid[m][n] < 0:
+                total += 1
+    return total
```
Comment (cclauss) [inline matrix/count_negative_numbers_in_sorted_matrix.py:105]:
```suggestion
    return len([number for row in grid for number in row if number < 0])
```

File: matrix/count_negative_numbers_in_sorted_matrix.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+    3
+    """
+    total = 0
+    length_of_n = len(grid[0])
+    for m in range(len(grid)):
+        for index, n in enumerate(range(length_of_n)):
+            if grid[m][n] < 0:
+                total += length_of_n - index
+                break
+    return total
```
Comment (cclauss) [inline matrix/count_negative_numbers_in_sorted_matrix.py:130]:
```suggestion
    total = 0
    for row in grid:
        for i, number in enumerate(row):
            if number < 0:
                total += len(row) - i
                break
    return total
```

File: matrix/count_negative_numbers_in_sorted_matrix.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+
+def generate_large_matrix() -> list[list[int]]:
+    """
+    >>> generate_large_matrix() # doctest: +ELLIPSIS
+    [[1000, ..., -999], [999, ..., -1001], ..., [2, ..., -1998]]
+    """
+    return [list(range(1000 - i, -1000 - i, -1)) for i in range(1000)]
+
+
```
Comment (cclauss) [inline matrix/count_negative_numbers_in_sorted_matrix.py:140]: Let's generate `grid` just once and then import it.
```suggestion

grid = generate_large_matrix()


```

File: matrix/count_negative_numbers_in_sorted_matrix.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+def benchmark() -> None:
+    """Benchmark our functions next to each other"""
+    from timeit import timeit
+
+    print("Running benchmarks")
+    setup = (
+        "from __main__ import count_negatives_binary_search,count_negatives_brute_force"
+        ",count_negatives_brute_force_with_break,generate_large_matrix"
```
Comment (cclauss) [inline matrix/count_negative_numbers_in_sorted_matrix.py:148]:
```suggestion
        ",count_negatives_brute_force_with_break,grid"
```

File: matrix/count_negative_numbers_in_sorted_matrix.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    )
+    for func in (
+        "count_negatives_binary_search",  # 175.51 seconds
+        "count_negatives_brute_force_with_break",  # 271.04 seconds
+        "count_negatives_brute_force",  # 646.65 seconds
+    ):
+        time = timeit(f"{func}(generate_large_matrix())", setup=setup, number=5000)
```
Comment (cclauss) [inline matrix/count_negative_numbers_in_sorted_matrix.py:155]: Let's keep the runtime for the fastest algorithm at about 1 second.
```suggestion
        time = timeit(f"{func}(grid=grid)", setup=setup, number=500)
```

**Review by cclauss — CHANGES_REQUESTED** [review]
This is looking cool!!

**Comment by CaedenPH** [comment]
@cclauss All changes have been implemented and tests are passing

**Review by cclauss — APPROVED** [review]
Nice!  Thanks for doing this.

**Comment by cclauss** [comment]
I have never figured out https://docs.python.org/3/library/bisect.html but it would be interesting to benchmark against an implementation that leveraged it.

**Pull request merged**